java.nio.charset.MalformedInputException in FileRegularExpressionCheck.java:120

I activated the FileRegularExpressionCheck for checking includes of stdafx.h in other header files but the analysis fails after some time. Looks like the encoding for the source files (*sonar.sourceEncoding*) is not handled in FileRegularExpressionCheck.java and the exception is not handled as well.
The analysis with sonar_runner fails:
```
ERROR: Error during Sonar runner execution
ERROR: Unable to execute Sonar
ERROR: Caused by: Unable to analyze file: C:\TFS_Server\Server\some-folder\some-header.h
ERROR: Caused by: java.nio.charset.MalformedInputException: Input length = 1
ERROR: Caused by: Input length = 1
``` 
The debug output has the details:
```
org.sonar.runner.impl.RunnerException: Unable to execute Sonar
        at org.sonar.runner.impl.BatchLauncher$1.delegateExecution(BatchLauncher.java:91)
        at org.sonar.runner.impl.BatchLauncher$1.run(BatchLauncher.java:75)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.sonar.runner.impl.BatchLauncher.doExecute(BatchLauncher.java:69)
        at org.sonar.runner.impl.BatchLauncher.execute(BatchLauncher.java:50)
        at org.sonar.runner.api.EmbeddedRunner.doExecute(EmbeddedRunner.java:102)
        at org.sonar.runner.api.Runner.execute(Runner.java:100)
        at org.sonar.runner.Main.executeTask(Main.java:70)
        at org.sonar.runner.Main.execute(Main.java:59)
        at org.sonar.runner.Main.main(Main.java:53)
Caused by: org.sonar.squidbridge.api.AnalysisException: Unable to analyze file: C:\TFS_Server\Server\some-folder\some-header.h
        at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:127)
        at org.sonar.plugins.cxx.squid.CxxSquidSensor.analyse(CxxSquidSensor.java:112)
        at org.sonar.batch.phases.SensorsExecutor.executeSensor(SensorsExecutor.java:79)
        at org.sonar.batch.phases.SensorsExecutor.execute(SensorsExecutor.java:70)
        at org.sonar.batch.phases.PhaseExecutor.execute(PhaseExecutor.java:119)
        at org.sonar.batch.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:194)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.scan.ProjectScanContainer.scan(ProjectScanContainer.java:233)
        at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:228)
        at org.sonar.batch.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:226)
        at org.sonar.batch.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:221)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.scan.ScanTask.scan(ScanTask.java:64)
        at org.sonar.batch.scan.ScanTask.execute(ScanTask.java:51)
        at org.sonar.batch.bootstrap.TaskContainer.doAfterStart(TaskContainer.java:125)
        at org.sonar.api.platform.ComponentContainer.startComponents(ComponentContainer.java:92)
        at org.sonar.api.platform.ComponentContainer.execute(ComponentContainer.java:77)
        at org.sonar.batch.bootstrap.BootstrapContainer.executeTask(BootstrapContainer.java:173)
        at org.sonar.batch.bootstrapper.Batch.executeTask(Batch.java:95)
        at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:67)
        at org.sonar.runner.batch.IsolatedLauncher.execute(IsolatedLauncher.java:48)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.sonar.runner.impl.BatchLauncher$1.delegateExecution(BatchLauncher.java:87)
        ... 9 more
Caused by: org.sonar.api.utils.SonarException: java.nio.charset.MalformedInputException: Input length = 1
        at org.sonar.cxx.checks.FileRegularExpressionCheck.visitFile(FileRegularExpressionCheck.java:100)
        at com.sonar.sslr.impl.ast.AstWalker.walkAndVisit(AstWalker.java:67)
        at org.sonar.squidbridge.AstScanner.scanFiles(AstScanner.java:106)
        ... 36 more
Caused by: java.nio.charset.MalformedInputException: Input length = 1
        at java.nio.charset.CoderResult.throwException(CoderResult.java:277)
        at java.nio.charset.CharsetDecoder.decode(CharsetDecoder.java:798)
        at org.sonar.cxx.checks.FileRegularExpressionCheck.fromFile(FileRegularExpressionCheck.java:120)
        at org.sonar.cxx.checks.FileRegularExpressionCheck.visitFile(FileRegularExpressionCheck.java:95)
        ... 38 more
```

--- a/cxx-checks/src/main/java/org/sonar/cxx/checks/FileRegularExpressionCheck.java
+++ b/cxx-checks/src/main/java/org/sonar/cxx/checks/FileRegularExpressionCheck.java
@@ -31,14 +31,14 @@ import org.sonar.squidbridge.checks.SquidCheck;
 
 import java.io.File;
 import java.io.FileInputStream;
-import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.CharBuffer;
 import java.nio.channels.FileChannel;
 import java.nio.charset.Charset;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CodingErrorAction;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
-import java.util.regex.PatternSyntaxException;
 import org.sonar.api.utils.PathUtils;
 import org.sonar.api.utils.WildcardPattern;
 
@@ -54,6 +54,7 @@ public class FileRegularExpressionCheck extends SquidCheck<Grammar> implements C
   private static final String DEFAULT_MESSAGE = "The regular expression matches this file";
 
   private Charset charset;
+  private CharsetDecoder decoder;
   private Pattern pattern;
 
   @RuleProperty(
@@ -75,7 +76,10 @@ public class FileRegularExpressionCheck extends SquidCheck<Grammar> implements C
   public void init() {
     try {
       pattern = Pattern.compile(regularExpression);
-    } catch (PatternSyntaxException e) {
+      decoder = charset.newDecoder();
+      decoder.onMalformedInput(CodingErrorAction.REPLACE);
+      decoder.onUnmappableCharacter(CodingErrorAction.REPLACE);
+    } catch (Exception e) {
       throw new SonarException(e);
     }
   }
@@ -92,11 +96,11 @@ public class FileRegularExpressionCheck extends SquidCheck<Grammar> implements C
         if (!matchFile()) {
           return;
         }
-        Matcher matcher = pattern.matcher(fromFile(getContext().getFile(), charset));
+        Matcher matcher = pattern.matcher(fromFile(getContext().getFile()));
         if (matcher.find()) {
           getContext().createFileViolation(this, message);
         }
-      } catch (IOException e) {
+      } catch (Exception e) {
         throw new SonarException(e);
       }
     }
@@ -111,19 +115,19 @@ public class FileRegularExpressionCheck extends SquidCheck<Grammar> implements C
     return true;
   }
 
-    private static CharSequence fromFile(File file, Charset charset) throws IOException {
-        FileInputStream input = null;
-        try {
-            input = new FileInputStream(file);
-            FileChannel channel = input.getChannel();
-            ByteBuffer bbuf = channel.map(FileChannel.MapMode.READ_ONLY, 0, (int) channel.size());
-            CharBuffer cbuf = charset.newDecoder().decode(bbuf);
-            return cbuf;
-        } finally {
-            if (input != null) {
-                input.close();
-            }
-        }
+  private CharSequence fromFile(File file) throws Exception {
+    FileInputStream input = null;
+    try {
+      input = new FileInputStream(file);
+      FileChannel channel = input.getChannel();
+      ByteBuffer bbuf = channel.map(FileChannel.MapMode.READ_ONLY, 0, (int) channel.size());
+      CharBuffer cbuf = decoder.decode(bbuf);
+      return cbuf;
+    } finally {
+      if (input != null) {
+        input.close();
+      }
     }
-    
+  }
+
 }

--- a/cxx-checks/src/test/java/org/sonar/cxx/checks/FileRegularExpressionCheckTest.java
+++ b/cxx-checks/src/test/java/org/sonar/cxx/checks/FileRegularExpressionCheckTest.java
@@ -20,9 +20,11 @@
 package org.sonar.cxx.checks;
 
 import java.io.File;
+import java.nio.charset.Charset;
 
 import org.junit.Test;
 import org.sonar.cxx.CxxAstScanner;
+import org.sonar.cxx.CxxConfiguration;
 import org.sonar.squidbridge.api.SourceFile;
 import org.sonar.squidbridge.checks.CheckMessagesVerifier;
 
@@ -31,10 +33,26 @@ public class FileRegularExpressionCheckTest {
   @Test
   public void fileRegExWithoutFilePattern() {
     FileRegularExpressionCheck check = new FileRegularExpressionCheck();
+    Charset charset = Charset.forName("UTF-8");
+    CxxConfiguration cxxConfig = new CxxConfiguration(charset);
     check.regularExpression = "stdafx\\.h";
     check.message = "Found 'stdafx.h' in file!";
 
-    SourceFile file = CxxAstScanner.scanSingleFile(new File("src/test/resources/checks/FileRegEx.cc"), check);
+    SourceFile file = CxxAstScanner.scanSingleFileConfig(new File("src/test/resources/checks/FileRegEx.cc"), cxxConfig, check);
+    CheckMessagesVerifier.verify(file.getCheckMessages())
+      .next().withMessage(check.message)
+      .noMore();
+  }
+
+  @Test
+  public void fileRegExCodingErrorActionReplace() {
+    FileRegularExpressionCheck check = new FileRegularExpressionCheck();
+    Charset charset = Charset.forName("US-ASCII");
+    CxxConfiguration cxxConfig = new CxxConfiguration(charset);
+    check.regularExpression = "stdafx\\.h";
+    check.message = "Found 'stdafx.h' in file!";
+
+    SourceFile file = CxxAstScanner.scanSingleFileConfig(new File("src/test/resources/checks/FileRegEx.cc"), cxxConfig, check);
     CheckMessagesVerifier.verify(file.getCheckMessages())
       .next().withMessage(check.message)
       .noMore();
@@ -43,11 +61,13 @@ public class FileRegularExpressionCheckTest {
   @Test
   public void fileRegExWithFilePattern() {
     FileRegularExpressionCheck check = new FileRegularExpressionCheck();
+    Charset charset = Charset.forName("UTF-8");
+    CxxConfiguration cxxConfig = new CxxConfiguration(charset);
     check.matchFilePattern = "/**/*.cc"; // all files with .cc file extension
     check.regularExpression = "#include\\s+\"stdafx\\.h\"";
     check.message = "Found '#include \"stdafx.h\"' in a .cc file!";
 
-    SourceFile file = CxxAstScanner.scanSingleFile(new File("src/test/resources/checks/FileRegEx.cc"), check);
+    SourceFile file = CxxAstScanner.scanSingleFileConfig(new File("src/test/resources/checks/FileRegEx.cc"), cxxConfig, check);
     CheckMessagesVerifier.verify(file.getCheckMessages())
       .next().withMessage(check.message)
       .noMore();

--- a/cxx-checks/src/test/resources/checks/FileRegEx.cc
+++ b/cxx-checks/src/test/resources/checks/FileRegEx.cc
@@ -1,4 +1,4 @@
-// Comment
+// Comment äöüß Copyright ©
    #include      "stdafx.h"
 int i;
 void func() {}

